Update materialized view configuration for redshift (#4589)
fishtown-analytics/docs.getdbt.com: website/docs/reference/resource-configs/redshift-configs.md

version: 2

models:
  - name: [<model-name>]
    config:
      [materialized](/reference/resource-configs/materialized): materialized_view
      on_configuration_change: apply | continue | fail
      [dist](#using-sortkey-and-distkey): all | auto | even | <field-name>
      [sort](#using-sortkey-and-distkey): <field-name> | [<field-name>]
      [sort_type](#using-sortkey-and-distkey): auto | compound | interleaved
      [auto_refresh](#auto-refresh): true | false
      [backup](#backup): true | false
```

</File>

</TabItem>

<TabItem value="config">

<File name='models/<model_name>.sql'>

```jinja
{{ config(
    [materialized](/reference/resource-configs/materialized)="materialized_view",
    on_configuration_change="apply" | "continue" | "fail",
    [dist](#using-sortkey-and-distkey)="all" | "auto" | "even" | "<field-name>",
    [sort](#using-sortkey-and-distkey)=["<field-name>"],
    [sort_type](#using-sortkey-and-distkey)="auto" | "compound" | "interleaved",
    [auto_refresh](#auto-refresh)=true | false,
    [backup](#backup)=true | false,
) }}
```

</File>

</TabItem>

</Tabs>

Many of these parameters correspond to their table counterparts and have been linked above.

> The parameters unique to materialized views are the auto-refresh and backup functionality, which are covered below.

Reviewer: ```suggestion
The parameters unique to materialized views are the [auto-refresh](#auto-refresh) and [backup](#backup) functionality, which are covered below.
```

Find more information about the [CREATE MATERIALIZED VIEW](https://docs.aws.amazon.com/redshift/latest/dg/materialized-view-create-sql-command.html) parameters in the Redshift docs.

#### Auto-refresh

| Parameter      | Type        | Required | Default | Change Monitoring Support |
|----------------|-------------|----------|---------|---------------------------|
| `auto_refresh` | `<boolean>` | no       | `false` | alter                     |

Redshift supports [automatic refresh](https://docs.aws.amazon.com/redshift/latest/dg/materialized-view-refresh.html#materialized-view-auto-refresh) configuration for materialized views.
By default, a materialized view does not automatically refresh.
dbt monitors this parameter for changes and applies them using an `ALTER` statement.

> Find more information about the [parameters](https://docs.aws.amazon.com/redshift/latest/dg/materialized-view-create-sql-command.html#mv_CREATE_MATERIALIZED_VIEW-parameters) in the Redshift docs.

Reviewer: ```suggestion
Learn more information about the [parameters](https://docs.aws.amazon.com/redshift/latest/dg/materialized-view-create-sql-command.html#mv_CREATE_MATERIALIZED_VIEW-parameters) in the Redshift docs.
```

#### Backup

| Parameter | Type        | Required | Default | Change Monitoring Support |
|-----------|-------------|----------|---------|---------------------------|
| `backup`  | `<boolean>` | no       | `true`  | n/a                       |

Redshift supports [backup](https://docs.aws.amazon.com/redshift/latest/mgmt/working-with-snapshots.html) configuration of clusters at the object level.
This parameter identifies if the materialized view should be backed up as part of the cluster snapshot.
By default, a materialized view will be backed up during a cluster snapshot.
dbt cannot monitor this parameter as it is not queryable within Redshift.
If the value is changed, the materialized view will need to go through a `--full-refresh` in order to set it.

> Find more information about the [parameters](https://docs.aws.amazon.com/redshift/latest/dg/materialized-view-create-sql-command.html#mv_CREATE_MATERIALIZED_VIEW-parameters) in the Redshift docs.

Reviewer: ```suggestion
Learn more about the [parameters](https://docs.aws.amazon.com/redshift/latest/dg/materialized-view-create-sql-command.html#mv_CREATE_MATERIALIZED_VIEW-parameters) in the Redshift docs.
```

## Materialized views

The Redshift adapter supports [materialized views](https://docs.aws.amazon.com/redshift/latest/dg/materialized-view-overview.html)
with the following configuration parameters:

| Parameter                                 | Type         | Required | Default              | Change Monitoring Support |
|-------------------------------------------|--------------|----------|----------------------|---------------------------|
| `on_configuration_change`                 | `<string>`   | no       | `apply`              | n/a                       |
| [`dist`](#using-sortkey-and-distkey)      | `<string>`   | no       | `even`               | drop/create               |
| [`sort`](#using-sortkey-and-distkey)      | `[<string>]` | no       | `none`               | drop/create               |
| [`sort_type`](#using-sortkey-and-distkey) | `<string>`   | no       | `auto` if no `sort`  | drop/create               |

> |                                           |              |          | `compound` if `sort` |                           |

Reviewer 2: @mikealfare are you intending a [mostly blank row](https://docs-getdbt-com-git-content-materialized-viewsredshift-dbt-labs.vercel.app/reference/resource-configs/redshift-configs#materialized-views) for the `compound` if `sort` portion? Or were you wanting it to be on the same row as `sort_type` to appear immediate after `auto` if no `sort`?

<img width="763" alt="image" src="https://github.com/dbt-labs/docs.getdbt.com/assets/44704949/b9009b8d-9f8d-47d6-b55f-6c396e1d1db2">

If the latter, you can add a `<br />` (or two) to separate those two different cases within the same cell or rewrite it slightly like this

| Parameter                                 | Type         | Required | Default              | Change Monitoring Support | 	
|-------------------------------------------|--------------|----------|----------------------|---------------------------|
| `on_configuration_change`                 | `<string>`   | no       | `apply`              | n/a                       |
| [`dist`](#using-sortkey-and-distkey)      | `<string>`   | no       | `even`               | drop/create               |
| [`sort`](#using-sortkey-and-distkey)      | `[<string>]` | no       | `none`               | drop/create               |
| [`sort_type`](#using-sortkey-and-distkey) | `<string>`   | no       | `auto` if not `sort` (`compound` otherwise)    | drop/create               |
| [`auto_refresh`](#auto-refresh)           | `<boolean>`  | no       | `false`              | alter                     |
| [`backup`](#backup)                       | `<string>`   | no       | `true`               | n/a                       |

Author: I was having a brain fart about that. Yes, that's what I wanted to do. I updated it.

|---------------------------|--------------|----------|------------------------|---------------------------|-------------------------------------------------|
| `on_configuration_change` | STRING       | NO       | `'apply'`              | N/A                       |                                                 |
| `dist`                    | STRING       | NO       | `'EVEN'`               | DROP/CREATE               | [Sortkey / Distkey](#using-sortkey-and-distkey) |
| `sort`                    | LIST[STRING] | NO       |                        | DROP/CREATE               | [Sortkey / Distkey](#using-sortkey-and-distkey) |
| `sort_type`               | STRING       | NO       | `'AUTO'` if no `sort`  | DROP/CREATE               | [Sortkey / Distkey](#using-sortkey-and-distkey) |
|                           |              |          | `'COMPOUND'` if `sort` |                           |                                                 |
| `auto_refresh`            | BOOLEAN      | NO       | `False`                | ALTER                     | [Auto refresh](#auto-refresh)                   |

> | `backup`                  | BOOLEAN      | HO       | `True`                 | N/A                       | [Backup](#backup)                               |

Reviewer 2: The table can be one column thinner if doing a hyperlink like [`dist`](#using-sortkey-and-distkey) in the first column instead.

Author: done

#### Sample model file:

<File name='redshift_materialized_view.sql'>

```sql
{{ config(
    materialized='materialized_view',
    on_configuration_change='{ apply | continue | fail }'

>     dist='{ ALL | AUTO | EVEN | <field_name> }',
>     sort=['<field_name>', ...],
>     sort_type='{ AUTO | COMPOUND | INTERLEAVED }'

Reviewer 2: Lowercase to match the casing in other examples?
```suggestion
    dist='{ all | auto | even | <field_name> }',
    sort=['<field_name>', ...],
    sort_type='{ auto | compound | interleaved }'
```

Author: I was matching the casing in the Redshift docs, though not doing so specifically. I think as far as functionality is concerned, both work (we're largely case-insensitive). And we're not necessarily tied to the platform docs as we regularly mess with configuration before submitting it to the database anyway. In other words, I have no strong feeling either way. Do you know off-hand what the style guide would say?

Author: I made them all lowercase to match what I saw elsewhere in the docs

| `sort_type`               | STRING       | NO       | `'AUTO'` if no `sort`  | DROP/CREATE               | [Sortkey / Distkey](#using-sortkey-and-distkey) |
|                           |              |          | `'COMPOUND'` if `sort` |                           |                                                 |
| `auto_refresh`            | BOOLEAN      | NO       | `False`                | ALTER                     | [Auto refresh](#auto-refresh)                   |
| `backup`                  | BOOLEAN      | HO       | `True`                 | N/A                       | [Backup](#backup)                               |

#### Sample model file:

<File name='redshift_materialized_view.sql'>

```sql
{{ config(
    materialized='materialized_view',
    on_configuration_change='{ apply | continue | fail }'
    dist='{ ALL | AUTO | EVEN | <field_name> }',
    sort=['<field_name>', ...],
    sort_type='{ AUTO | COMPOUND | INTERLEAVED }'
    auto_refresh=<bool>,
    backup=<bool>,
) }}

select * from {{ ref('my_base_table') }}
```

</File>

Many of these parameters correspond to their table counterparts and have been linked above.

> The set of parameters which are unique to materialized views covers auto-refresh and backup functionality, which is covered below.

Reviewer: ```suggestion
The parameters unique to materialized views are the auto-refresh and backup functionality, which are covered below.
```

> Find more information about these parameters in the Redshift docs:

Reviewer: ```suggestion
Find more information about the [CREATE MATERIALIZED VIEW](https://docs.aws.amazon.com/redshift/latest/dg/materialized-view-create-sql-command.html) parameters in the Redshift docs.
```

> - [CREATE MATERIALIZED VIEW](https://docs.aws.amazon.com/redshift/latest/dg/materialized-view-create-sql-command.html)

Reviewer: ```suggestion

```

#### Auto-refresh

| Parameter      | Type    | Required | Default | Change Monitoring Support |
|----------------|---------|----------|---------|---------------------------|
| `auto_refresh` | BOOLEAN | NO       | `False` | ALTER                     |

Redshift supports [automatic refresh](https://docs.aws.amazon.com/redshift/latest/dg/materialized-view-refresh.html#materialized-view-auto-refresh) configuration for materialized views.

> By default, a materialized view will not automatically refresh.

Reviewer: ```suggestion
By default, a materialized view does not automatically refresh.
```

> dbt will monitor this parameter for changes and apply them using an `ALTER` statement.

Reviewer: ```suggestion
dbt monitors this parameter for changes and applies them using an `ALTER` statement.
```

> Find more information about this parameter in the Redshift docs:

Reviewer: ```suggestion
Find more information about the [parameters](https://docs.aws.amazon.com/redshift/latest/dg/materialized-view-create-sql-command.html#mv_CREATE_MATERIALIZED_VIEW-parameters) in the Redshift docs.
```

> - [Parameters](https://docs.aws.amazon.com/redshift/latest/dg/materialized-view-create-sql-command.html#mv_CREATE_MATERIALIZED_VIEW-parameters)

Reviewer: ```suggestion

```

#### Backup

| Parameter | Type    | Required | Default | Change Monitoring Support |
|-----------|---------|----------|---------|---------------------------|
| `backup`  | BOOLEAN | NO       | `True`  | N/A                       |

Redshift supports [backup](https://docs.aws.amazon.com/redshift/latest/mgmt/working-with-snapshots.html) configuration of clusters at the object level.
This parameter identifies if the materialized view should be backed up as part of the cluster snapshot.
By default, a materialized view will be backed up during a cluster snapshot.
dbt cannot monitor this parameter as it is not queryable within Redshift.
If the value is changed, the materialized view will need to go through a `--full-refresh` in order to set it.

> Find more information about this parameter in the Redshift docs:

Reviewer: ```suggestion
Find more information about the [parameters](https://docs.aws.amazon.com/redshift/latest/dg/materialized-view-create-sql-command.html#mv_CREATE_MATERIALIZED_VIEW-parameters) in the Redshift docs.
```

> - [Parameters](https://docs.aws.amazon.com/redshift/latest/dg/materialized-view-create-sql-command.html#mv_CREATE_MATERIALIZED_VIEW-parameters)

Reviewer: ```suggestion

```

### Limitations

As with most data platforms, there are limitations associated with materialized views. Some worth noting include:

> - Materialized views cannot reference: views, temporary tables, user defined functions, late-binding tables

Reviewer: ```suggestion
- Materialized views cannot reference views, temporary tables, user-defined functions, or late-binding tables.
```

> - Auto-refresh cannot be used if the materialized view references: mutable functions, external schemas, another materialized view

Reviewer: ```suggestion
- Auto-refresh cannot be used if the materialized view references mutable functions, external schemas, or another materialized view.
```
